blog/src/a-base32-checksum.md

```diff
@@ -0,0 +1,1241 @@
+    for (let j = 0; j < width; j++) {
+      if ((a[i] & (1 << (width - j - 1))) !== 0) {
+        mat[i] ^= b[j];
+      }
+    }
+  }
+  return mat;
+}
+
+function matCp(a) {
+  let copy = new Array(a.length);
+  for (let i = 0; i < a.length; i++) {
+    copy[i] = a[i];
+  }
+  return copy;
+}
+
+const primitivePowers = (function genPowersOfPrimitive() {
+  // Index 0 contains P^0 = I, 1 has P^1, … 30 has P^30.
+  const powers = [ [], matCp(primitive) ];
+  let p = powers[1];
+  for (let i = 0; i < cardinal - 3; i++) {
+    p = matMul(p, primitive);
+    powers.push(p);
+  }
+  powers[0] = matMul(p, primitive);
+  return powers;
+})();
+
+function fromBase32Char(c) {
+  return c.charCodeAt(0) - (/[a-z]/.test(c)? 97: 24);
+}
+
+function toBase32Char(c) {
+  c = +c;
+  const d = (c > 25)? (c + 24): (c + 97);
+  return String.fromCharCode(d);
+}
+
+function compute(payload) {
+  payload = String(payload).toLowerCase();
+  const n = payload.length;
+
+  // We must solve Σ ai P^i = 0 for i from 1 to n+1.
+  // First, compute Σ ai P^i for i from 1 to n.
+  let sum = 0;
+  for (let i = 0; i < n; i++) {
+    const a = fromBase32Char(payload[i]);
+    sum ^= matMul([a], primitivePowers[(i+1) % (cardinal-1)])[0];
+    //console.log(`a ${a}\texp ${i+1}\tsum ${sum}`);
+  }
+
+  // We must solve:  sum + code * primitive^(n+1) = 0
+  // That is:        sum + opposite           = 0
+  // Therefore:      opposite = -sum
+  // In GF(2), matrices are their own opposites.
+  const opposite = sum;
+
+  // We must solve:  code * primitive^(n+1) = opposite
+  // We know:        a^(cardinal-1) = 1
+  // Therefore:      a * a^(cardinal-2) = 1
+  // Here we have:   a = primitive^(n+1)
+  // Hence:          code = opposite * primitive^((cardinal-2)*(n+1))
+  let exp = (cardinal-n-2) % (cardinal - 1);
+  exp = (exp < 0)? exp + cardinal: exp;
+  const inverse = primitivePowers[exp];
+  const code = matMul([opposite], inverse)[0];
+  //console.log(`opposite ${opposite}\texp ${exp}\tinverse ${JSON.stringify(inverse)}\tcode ${code}`);
+  return toBase32Char(code);
+}
+
+function validate(payload) {
+  return compute(payload) === 'a';
+}
+
+function init() {
+  const widget = document.getElementById('base32check1Widget');
+  const input = widget.getElementsByClassName('input')[0];
+  const validity = widget.getElementsByClassName('validity')[0];
+  const checksum = widget.getElementsByClassName('checksum')[0];
+  const updateWidget = () => {
+    const value = input.value;
+    validity.textContent = validate(value) ?  'Valid' : 'Invalid';
+    const check = compute(value);
+    checksum.textContent = (check !== check) ? 'removal of non-base32' : check;
+  };
+  input.addEventListener('input', updateWidget);
+  updateWidget();
+}
+addEventListener('DOMContentLoaded', init);
+})();
+</script>
+
+## base32check2
+
+I chose to also design a 2-character checksum alongside, on the off-chance that
+the first was not strong enough. There are 2<sup>2×5</sup> = 1024 possible
```

Comment:
What makes you doubt that the first isn't strong enough?

Reply:
I am making a single (5-bit) character check compete with a double character check (6.6 bits total). I needed to give better results with less space. Even though it ended up working out, it was a bit ambitious.

I will clarify this initial estimate.